File: Year02/2nd semester/MPP/Lab/mpp-proiect-java-AlexandruPaul21/JavaProject/src/main/java/org/example/repository/AgencyRepository.java
package org.example.repository;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.example.model.Agency;
import org.example.utils.JdbcUtils;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

public class AgencyRepository implements IAgencyRepository {

    private final JdbcUtils jdbcUtils;
    private static final Logger logger = LogManager.getLogger();

    public AgencyRepository(Properties properties) {
        logger.info("Initialising Agency Repository with properties: {}", properties);
        jdbcUtils = new JdbcUtils(properties);
    }

    @Override
    public Agency findOne(String s) {
        logger.traceEntry();

        String sql = "SELECT * FROM agencies WHERE username = ?";
        Agency agency = null;
        try (
                Connection connection = jdbcUtils.getConnection()
                ) {
            PreparedStatement preparedStatement = connection.prepareStatement(sql);
            preparedStatement.setString(1, s);

            ResultSet resultSet = preparedStatement.executeQuery();

            if (resultSet.next()) {
                String username = resultSet.getString("username");
                String agName = resultSet.getString("ag_name");
                String password = resultSet.getString("pass");

                agency = new Agency(username, agName, password);
            }
        } catch (SQLException e) {
            logger.error(e);
            e.printStackTrace();
        }
        logger.traceExit();
        return agency;
    }

    @Override
    public Iterable<Agency> findAll() {
        logger.traceEntry();

        List<Agency> agencies = new ArrayList<>();

        String sql = "SELECT * FROM agencies";
        try (
                Connection connection = jdbcUtils.getConnection()
        ) {
            PreparedStatement preparedStatement = connection.prepareStatement(sql);
            ResultSet resultSet = preparedStatement.executeQuery();

            while (resultSet.next()) {
                String username = resultSet.getString("username");
                String agName = resultSet.getString("ag_name");
                String password = resultSet.getString("pass");

                Agency agency = new Agency(username, agName, password);

                agencies.add(agency);
            }
        } catch (SQLException e) {
            logger.error(e);
            e.printStackTrace();
        }
        logger.traceExit();
        return agencies;
    }

    @Override
    public Agency save(Agency entity) {
        logger.traceEntry();
        String sql = "INSERT INTO agencies VALUES (?, ?, ?)";

        try (
                Connection connection = jdbcUtils.getConnection()
        ) {
            PreparedStatement preparedStatement = connection.prepareStatement(sql);
            preparedStatement.setString(1, entity.getId());
            preparedStatement.setString(2, entity.getName());
            preparedStatement.setString(3, entity.getPassword());

            preparedStatement.executeUpdate();
            entity = null;
        } catch (SQLException e) {
            logger.error(e);
            e.printStackTrace();
        }
        logger.traceExit();
        return entity;
    }
}

File: Year02/2nd semester/MPP/Lab/mpp-proiect-csharp-AlexandruPaul21/CsharpProj/repository/AgencyRepository.cs
using System.Configuration;
using CsharpProj.model;
using log4net;
using Npgsql;
// ReSharper disable All

namespace CsharpProj.repository;

public class AgencyRepository : IAgencyRepository
{
    //private readonly Logger logger = new Logger();
    
    private static readonly ILog logger = LogManager.GetLogger("AgencyRepository");

    public Agency findOne(string Id)
    {
        logger.Info("Agency - findOne - Enter");
        Agency agency = new Agency("","");
        using (var connection = new NpgsqlConnection(ConfigurationManager.ConnectionStrings["Test"].ConnectionString))
        {
            connection.Open();

            var sql = "SElECT * FROM agencies WHERE username = @username";
            using (var cmd = new NpgsqlCommand(sql, connection))
            {
                cmd.Parameters.AddWithValue("username", Id);
                var reader = cmd.ExecuteReader();
                if (reader.Read())
                {
                    string username = reader.GetString(0);
                    string name = reader.GetString(1);
                    string password = reader.GetString(2);

                    agency = new Agency(name, password);
                    agency.Id = username;
                }
            }
        }
        logger.Info("Agency - findOne - Exit");
        return agency;
    }

    public IEnumerable<Agency> findAll()
    {
        logger.Info("Agency - findAll - Enter");
        List<Agency> agencies = new List<Agency>();
        using (var connection = new NpgsqlConnection(ConfigurationManager.ConnectionStrings["Test"].ConnectionString))
        {
            connection.Open();

            var sql = "SElECT * FROM agencies";
            using (var cmd = new NpgsqlCommand(sql, connection))
            {
                var reader = cmd.ExecuteReader();
                while (reader.Read())
                {
                    string username = reader.GetString(0);
                    string name = reader.GetString(1);
                    string password = reader.GetString(2);

                    Agency agency = new Agency(name, password);
                    agency.Id = username;
                    
                    agencies.Add(agency);
                }
            }
        }
        logger.Info("Agency - findAll - Exit");
        return agencies;
    }

    public Agency save(Agency entity)
    {
        logger.Info("Agency - save - Enter");
        using (var connection = new NpgsqlConnection(ConfigurationManager.ConnectionStrings["Test"].ConnectionString))
        {
            connection.Open();

            var sql = "INSERT INTO agencies VALUES (@username, @name, @password)";
            using (var cmd = new NpgsqlCommand(sql, connection))
            {
                cmd.Parameters.AddWithValue("@username", entity.Id);
                cmd.Parameters.AddWithValue("@name", entity.Name);
                cmd.Parameters.AddWithValue("@password", entity.Password);

                int rowsAff = cmd.ExecuteNonQuery();
                if (rowsAff != 0)
                {
                    entity = null!;
                }
            }
        }
        logger.Info("Agency - save - Exit");
        return entity;
    }

    public bool existsUser(string username, string password)
    {
        logger.Info("Agency - existsUser - Enter");
        bool exists = false;
        using (var connection = new NpgsqlConnection(ConfigurationManager.ConnectionStrings["Test"].ConnectionString))
        {
            connection.Open();

            var sql = "SELECT * FROM agencies WHERE username = @username AND pass = @password";
            using (var cmd = new NpgsqlCommand(sql, connection))
            {
                cmd.Parameters.AddWithValue("@username", username);
                cmd.Parameters.AddWithValue("@password", password);

                var reader = cmd.ExecuteReader();
                if (reader.Read())
                {
                    exists = true;
                }
            }
        }
        logger.Info("Agency - existsUser - Exit");
        return exists;
    }
}